bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,325 @@
+            gloss_id,
+            video_id,
+            gloss_public.lower() == "t",
+            video_public.lower() == "t",
+        ]
+
+    print(
+        f"{len(this_nzsl_raw_keys_dict)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_nzsl_raw_keys_dict
+
+
+# Get all keys from AWS S3
+def get_s3_bucket_raw_keys_list(s3_bucket=AWS_S3_BUCKET):
+    print(f"Getting raw AWS S3 keys recursively ({s3_bucket}) ...", file=sys.stderr)
+
+    s3_resource = boto3.resource("s3")
+    s3_resource_bucket = s3_resource.Bucket(s3_bucket)
+    this_s3_bucket_raw_keys_list = [
+        s3_object.key for s3_object in s3_resource_bucket.objects.all()
+    ]
+
+    print(
+        f"{len(this_s3_bucket_raw_keys_list)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_s3_bucket_raw_keys_list
+
+
+# Get the keys present and absent across NZSL Signbank and S3, to dictionary
+def create_all_keys_dict(this_nzsl_raw_keys_dict, this_s3_bucket_raw_keys_list):
+    print(
+        "Getting keys present and absent across NZSL Signbank and S3 ...",
+        file=sys.stderr,
+    )
+    this_all_keys_dict = {}
+
+    # Find S3 keys that are present in NZSL, or absent
+    for video_key in this_s3_bucket_raw_keys_list:
+        dict_row = this_nzsl_raw_keys_dict.get(video_key, None)
+        if dict_row:
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                True,  # S3 PRESENT
+            ] + dict_row
+        else:
+            this_all_keys_dict[video_key] = [
+                False,  # NZSL Absent
+                True,  # S3 PRESENT
+            ] + [""] * 6
+
+    # Find NZSL keys that are absent from S3 (present handled above)
+    for video_key, dict_row in this_nzsl_raw_keys_dict.items():
+        if video_key not in this_s3_bucket_raw_keys_list:
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+            ] + dict_row
+
+    return this_all_keys_dict
+
+
+# Cases
+# In S3     In NZSL     Action
+#   Is        Not         Delete S3 Object
+#   Is        Is          Update ACL
+#   Not       Is          Review
+#      Other              Review
+def get_recommended_action(key_in_nzsl, key_in_s3):
+    if key_in_s3:
+        if key_in_nzsl:
+            return "Update ACL"
+        else:
+            return "Delete S3 Object"
+    return "Review"
+
+
+# Get S3 object's ACL
+def get_s3_canned_acl(video_key):
+    # TODO pass in a boto client instead of recreating one each time
+    s3_client = boto3.client("s3")
+    acls_grants = s3_client.get_object_acl(Bucket=AWS_S3_BUCKET, Key=video_key)[
+        "Grants"
+    ]
+    if len(acls_grants) > 1:
+        if (
+            acls_grants[0]["Permission"] == "FULL_CONTROL"
+            and acls_grants[1]["Permission"] == "READ"
+        ):
+            return "public-read"
+    elif acls_grants[0]["Permission"] == "FULL_CONTROL":
+        return "private"
```

Comment:
Thought: If you extracted all the "Permissions" into a set you could perhaps do some dictionary lookups instead of nested ifs

```python

grant_sets = {
    set(["READ", "FULL_CONTROL"]) = "public-read"
    set(["FULL_CONTROL"]) = "private"
}
permissions = set([g["Permissions"] for g in acls_grants))

return grant_sets.get(permissions, "unknown")
```

This does depend on what other values and combinations there might be. If it does get more complex you could always invert the dictionary, loop through them and use set comparisons so assess coverage of permissions.

Reply:
To my shame, I had no idea sets could be used as dictionary keys! Though I believe they have to be rendered immutable somehow?
Anyway, nice idea because it makes the matching data driven rather than conditional driven, but I'll skip it for now for time.